Issue #5264 Demo modules

gregw opened the pull request:
Create modules in jetty-home to download and deploy demos for #5264

gregw committed:
Issue #5264 Demo modules

Create modules in jetty-home to download and deploy demos

gregw commented:
@sbordet Do you like this approach?

joakime commented:
This could be done easier without including everything and the kitchen sink in jetty-home itself.
Why can't we use / unpack the `config` classified artifacts here? (why are we including the various xml that belongs to the various demos in jetty-home itself?)

gregw commented:
@joakime this is a draft, just to see if we like how the approach works.
If the basic approach is OK, then yes assembling it from configs would be better. For now I didn't want to  break those configs and thus break distro, as I am using it as a comparison.

gregw commented:
@joakime I'm in and out mostly for the next few days... so as this is on critical path for the release, can I pass it over to you?
@sbordet agrees with the basic approach as it lines up with his doco efforts well.  Thus what is left to do includes:
 + build a few more modules to make --add-module=demo-base produce more or less identical results to what we already have in demo-base.  I know I'm missing jaas and jndi webapps... but probably a few more things.
 + assemble those modules into jetty-home using the config mechanism - probably moving the mod files etc. back out to the jetty-xxx modules...
 + make sure that all the artifacts we get from maven will be published with a release.
 + either get rid of jetty-distribution... or better yet just make it use a `--add-module=demo-base` to create itself for now.

This will need lots of testing... so perhaps that's how I can help by tuning in and doing some testing and generally being a PITA.
If you can't take this hot potato, can you pass onto somebody else?

joakime committed:
Issue #5264 - Supporting snapshot maven repo metadata in start.jar

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 - Supporting extract of maven archive to destination

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime commented:
I've done a few things in the [jetty-10.0.x-5264-demo-with-start-extract](https://github.com/eclipse/jetty.project/compare/jetty-10.0.x-5264-demo-with-start-extract) branch.

Since I needed to test download from a repository, i added the following features to `jetty-start`.

* Now supports SNAPSHOT repositories properly (reads the `maven-metadata.xml` to figure out the snapshot version)
* Now supports extract on maven `[files]` section ... `maven://<groupId>/<artifactId>/<version>/<type>/<classfier>|extract:<path-in-jetty-base>`

This way we can use the `maven-deploy-plugin` to test a repository on the local disk (to know if we have everything properly deploying).

Example:
```
$ mkdir /tmp/jetty/snapshot-repo
$ cd /tmp/jetty/snapshot-repo
$ SNAPSHOT_REPO=`pwd`
$ cd /path/to/jetty.project-10.0.x
$ mvn clean install deploy -DskipTests -DaltSnapshotDeploymentRepository=jettysnapshots::default::file://$SNAPSHOTS_REPO
# and then later when testing ...
$ java -Dmaven.repo.uri=file://tmp/jetty/snapshot-repo -jar start.jar --add-to-start=demo-base
```

That branch also has a `home-config` assembly that will create `home` classified artifacts if the `src/main/home` directory exists.
This is so that the artifacts can overlay `jetty-home` specific configuration for the building of `jetty-home`.
That way we can keep the modules (and other files) in near their artifacts.

That branch also has a `/tests/test-webapps/demo-base-webapp` which is essentially the `webapps/ROOT` (courtesy of "-config.jar") and the `demo-base.mod` (courtesy of `-home.jar`)

I've started moving the demos files to their specific artifacts, but I'm not done.
I'll continue my efforts on monday.

gregw commented:
@joakime why do we need to create a new home config concept?   All configs should now go to jetty-home and none should go to jetty-distribution.  So we should be able to continue using src/main/config and don't need a src/main/home (even though that is a better name :)

joakime commented:
`src/main/config` is for the live jetty.base. (it's for any optional components, that don't ship with jetty-home, that are downloaded as part of a module)
`src/main/home` is for the jetty.home. (is for managing the jetty-home files, keeping the module, and it's relevant configs, near the artifact that being managed by source control)

gregw commented:
@joakime I don't get the distinction you are trying to make?    `config` is for files that get expanded into `jetty-home` during the build process, which will now include the demo modules.
I just don't think we want or need to introduce an extra mechanism at this point.    We know more or less how we want jetty-home to look - this PR shows that.  We just need to use the existing config mechanism to assemble it.

gregw commented:
@joakime so are you saying that instead of having some files in the `module/demo.d` file, you want to re-purpose the `config` mechanism to be published artefacts of config files that can be downloaded by modules and the existing `config` mechanism is being renamed to `home` and it is just used to assemble during a build?

Hmmmm I don't dislike it, but it just feels a bit too much to add at this very VERY late state, just to avoid a `module/demo.d` directory with a few xml and similar small files.  I'd much rather get this done with the features we have (although expand is a good addition).

janbartel commented:
IIUC @joakime this is a proposal to allow demo modules to be downloaded? I'm in favour of having them be downloaded. I'm not in favour of a system whereby I download a jetty distribution and there's nothing in it essentially other than a means to download everything: I want to be able to look at the modules and the xml files and investigate them before I enable any of them.

I'm also very concerned about this new circular dependency that has arisen: it should absolutely be possible to cd to tests/test-webapps/test-jetty-webapp and just type mvn jetty:run. If we can't do that then something is wrong.

gregw commented:
@joakime I think I understand the problem you are trying to solve.... it is a bit funny having the distributed jetty-home contain `module/demo-foo.mod` files and `modules/demo.d/demo-foo.xml` files, but then to download the war or libs used by the module.

But I don't think your solution helps as it breaks an optional module into 3 things:  1) the mod file baked into jetty-home; 2) the published artefact (war or lib); 3) the published config artefact

I better mid term solution would be to support remote modules.  Ie `demo-foo.mod` would not be part of jetty-home at all, nor any xml files.  Instead we could support `--add-module=http://some/url/to/the/demo-foo.module`, and this URL (or mvn coords) would download the `module/demo-foo.mod` files and `modules/demo.d/demo-foo.xml` files, maybe even the war and libs. This module could then be enabled, which might then download more files.

I think remote modules is a better solution to the problem you are trying to solve.... but too late for that feature.

Let's just go with a `modules/demo.d` directory containing a few xmls etc that is assembled with our current config mechanism.

joakime committed:
Issue #5264 - Reworking test-distribution to use jetty-home

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 - Disabling jetty-distribution from build reactor

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 - Reworking demo/tests into modules based on config classifiers

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 - Fixing deploy status on demo webapps

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 - Removing compact3 build

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 - Fixing demo-jndi config

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 - Fixing demo-jndi deps for javax.mail

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

gregw committed:
added jstl

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
added jndi and jaas dependencies

Signed-off-by: Greg Wilkins <gregw@webtide.com>

joakime committed:
Issue #5264 - Changing Distro to Home in jetty-maven-plugin

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

olamy committed:
fix maven plugin starting distribution as jetty home has different name, configure invoker plugin to be faster when running locally

Signed-off-by: olivier lamy <oliver.lamy@gmail.com>

olamy commented on jetty-start/src/test/java/org/eclipse/jetty/start/fileinits/MavenLocalRepoFileInitializerTest.java:
```diff
@@ -172,28 +177,90 @@ public void testGetCoordinateTestMavenBaseUri()
+        FS.ensureEmpty(snapshotLocalRepoDir);
+
+        MavenLocalRepoFileInitializer repo =
+            new MavenLocalRepoFileInitializer(baseHome, snapshotLocalRepoDir, false, "https://oss.sonatype.org/content/repositories/jetty-snapshots/");
```
I don't like much the idea about having test depending on downloading resources from https://oss.sonatype.org.
@joakime not urgent but I can fix that (next week when back from holidays)

joakime replied:
We could setup a small remote repo in `src/test/resources` (a 1 artifact repo) for testing purposes.

joakime replied:
I'll work on this today.

olamy commented on jetty-start/src/test/java/org/eclipse/jetty/start/fileinits/MavenMetadataTest.java:
```diff
@@ -0,0 +1,107 @@
+
+public class MavenMetadataTest
+{
+    private static final TimeZone UTC = TimeZone.getTimeZone("UTC");
```
variable not used

olamy commented on jetty-start/src/test/java/org/eclipse/jetty/start/fileinits/MavenMetadataTest.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    @Test
+    public void testIsExpiredTimestamp12HoursOld()
```
this test will fail is executed before noon :)

joakime replied:
Yeah, i caught that 1 failure. :-)
I'm thinking of just dropping that specific test method.

olamy commented on jetty-start/src/test/java/org/eclipse/jetty/start/fileinits/MavenMetadataTest.java:
```diff
@@ -0,0 +1,107 @@
+    @Test
+    public void testIsExpiredTimestamp12HoursOld()
+    {
+        LocalDateTime oldish = LocalDateTime.now().minusHours(12);
```
what about something such `now(ZoneId.of("UTC"))` as `getTimestampFormatter` use `UTC`?

joakime replied:
That might work.

joakime replied:
I opted to just remove this specific test, it's not reliable as written, and is too sensitive to the time of day that surefire user executes the test on.

olamy commented on jetty-start/src/main/java/org/eclipse/jetty/start/fileinits/MavenLocalRepoFileInitializer.java:
```diff
@@ -219,6 +290,65 @@ public Coordinates getCoordinates(URI uri)
+        try
+        {
+            MavenMetadata mavenMetadata = new MavenMetadata(localRepoMetadataPath);
+            return MavenMetadata.isExpiredTimestamp(mavenMetadata.getLastUpdated());
```
maybe we could have some configuration à la `mvn -U` to force downloading SNAPSHOT as here we have a sort of hardcoded value (the day before) maybe a sys property?

joakime replied:
Forcing download of new snapshots could be useful, yes.
Lets make that a new PR after this one gets merged.

joakime replied:
This ability, to support SNAPSHOT artifacts, is currently only useful to us when testing things.
The actual 10.0.0 release wouldn't have this requirement.
So I think it's safe to push this `--update-snapshots` feature to a future Jetty 10.0.x release.

joakime committed:
Issue #5264 - Removing 12 hour expired test (too time of day sensitive)

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 - Removing unused variable

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

gregw committed:
Merge branch 'jetty-10.0.x' into jetty-10.0.x-5264-demo-modules

sbordet commented on examples/async-rest/async-rest-webapp/src/main/config/modules/demo-async-rest.mod:
```diff
@@ -0,0 +1,15 @@
+deploy
+
+[files]
+maven://org.eclipse.jetty.example-async-rest/example-async-rest-webapp/${jetty.version}/war|webapps/demo-async-rest.war
```
The `groupId` should be `org.eclipse.jetty.examples` like we have `o.e.j.tests`.

joakime replied:
Done.

sbordet commented on jetty-home/pom.xml:
```diff
@@ -464,7 +464,7 @@
             </goals>
             <configuration>
-              <includeGroupIds>org.eclipse.jetty,org.eclipse.jetty.websocket</includeGroupIds>
+              <includeGroupIds>org.eclipse.jetty,org.eclipse.jetty.tests,org.eclipse.jetty.websocket</includeGroupIds>
```
Shouldn't the `async-rest` groupId be here too?

joakime replied:
Done.

sbordet commented on jetty-home/src/main/resources/modules/demo-base.mod:
```diff
@@ -0,0 +1,28 @@
+demo-async-rest
+demo-proxy
+demo-spec
+
```
Unnecessary empty line.

joakime replied:
Done.

gregw committed:
Merge remote-tracking branch 'origin/jetty-10.0.x' into jetty-10.0.x-5264-demo-modules

sbordet commented on jetty-start/src/main/java/org/eclipse/jetty/start/fileinits/MavenLocalRepoFileInitializer.java:
```diff
@@ -219,6 +290,65 @@ public Coordinates getCoordinates(URI uri)
+    protected void download(Coordinates coords, Path destination)
+        throws IOException
+    {
+        if (coords.version.endsWith("-SNAPSHOT"))
```
I don't think this is necessary at all - seems to introduce a lot of additional complexity.
Maybe it's a generic feature but I don't think it should go into this PR.

joakime replied:
As discussed, this is ok.

gregw committed:
demo-jaas depends on jaas

Signed-off-by: Greg Wilkins <gregw@webtide.com>

sbordet commented on tests/test-webapps/test-jaas-webapp/pom.xml:
```diff
@@ -18,50 +18,37 @@
-          <skip>true</skip>
+          <!-- DEMO Project, Must Deploy -->
+          <skip>false</skip>
         </configuration>
```
Just remove the `<configuration>` section altogether.

sbordet commented on tests/test-webapps/test-jetty-webapp/pom.xml:
```diff
@@ -88,50 +86,55 @@
-          <!-- must deploy: required for jetty-distribution -->
+          <!-- DEMO Project, Must Deploy -->
           <skip>false</skip>
         </configuration>
```
Remove `<configuration>` section.

sbordet commented on tests/test-webapps/test-jetty-webapp/pom.xml:
```diff
@@ -88,50 +86,55 @@
+        </plugin>
+
+      </plugins>
+    </pluginManagement>
```
Why has this part been moved from `<build>` to `<pluginManagement>`? We don't need to manage plugins for children POMs, no?

joakime replied:
Having jetty-maven-plugin as a formal plugin declaration forces it to resolve in the maven reactor.
It was causing a cyclic dependency as  ...

```
        WARNING: not enabled by default, as it introduces a cyclic reference in the build reactor.
        [ERROR] The projects in the reactor contain a cyclic reference:
          Edge between 'Vertex{label='org.eclipse.jetty:jetty-home:10.0.0-SNAPSHOT'}' and
          'Vertex{label='org.eclipse.jetty:test-jetty-webapp:10.0.0-SNAPSHOT'}'
          introduces to cycle in the graph:
           1: org.eclipse.jetty:test-jetty-webapp:10.0.0-SNAPSHOT
           2: org.eclipse.jetty:jetty-maven-plugin:10.0.0-SNAPSHOT
           3: org.eclipse.jetty:jetty-home:10.0.0-SNAPSHOT
           4: org.eclipse.jetty:test-jetty-webapp:10.0.0-SNAPSHOT @
```

Simply moving it to `<pluginManagement>` eliminates it from being involved in the reactor, but still allows it to be used from the command line at will.

sbordet commented on tests/test-webapps/test-jetty-webapp/src/main/config/modules/demo.d/demo-jetty-override-web.xml:
```diff
@@ -0,0 +1,64 @@
+   xmlns="http://xmlns.jcp.org/xml/ns/javaee" 
+   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+   xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/web-app_3_1.xsd" 
+   version="3.1"> 
```
Use 4.0.

joakime replied:
Oddly, no other web descriptor in `jetty-10.0.x` uses the correct XML intro.

From: https://javaee.github.io/servlet-spec/downloads/servlet-4.0/servlet-4_0_FINAL.pdf

``` xml
<web-app xmlns="http://xmlns.jcp.org/xml/ns/javaee"
xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee webapp_4_0.xsd"
version="4.0">
```

Changing this breaks things.
Lets work this out in a different PR.

joakime replied:
Opened as Issue #5319

sbordet commented on tests/test-webapps/test-jndi-webapp/pom.xml:
```diff
@@ -18,8 +18,8 @@
-          <skip>true</skip>
+          <!-- DEMO Project, Must Deploy -->
+          <skip>false</skip>
         </configuration>
```
Remove `<configuration>` section.

sbordet commented on tests/test-webapps/test-jndi-webapp/src/main/config/modules/demo.d/demo-jndi.xml:
```diff
@@ -0,0 +1,74 @@
+<?xml version="1.0"?><!DOCTYPE Configure PUBLIC "-//Jetty//Configure//EN" "https://www.eclipse.org/jetty/configure_9_3.dtd">
```
Must be `configure_10_0.dtd`.

joakime replied:
Fixed in 1feb60bb2c8e521f80aaee2abc901c067b561026 (for many files in `/demos/`)

sbordet commented on tests/test-webapps/test-servlet-spec/test-spec-webapp/pom.xml:
```diff
@@ -18,7 +18,8 @@
-          <skip>true</skip>
+          <!-- DEMO Project, Must Deploy -->
+          <skip>false</skip>
         </configuration>
```
Remove `<configuration>`

joakime replied:
Fixed

sbordet reviewed (changes requested):
Seems there are too many things in this PR - I thought it was just a couple of modules added?

gregw committed:
demo-jaas does not depend on demo-jndi

Signed-off-by: Greg Wilkins <gregw@webtide.com>

joakime committed:
Issue #5264 - Moving demo webapps to /demos/ tree

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 - Updating embedded JettyDistro to JettyHome

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 - Updating Configure DTD to 10_0 for Demos

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 - Adding demo.mod support to embedded examples tests

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 - Work around junit-vintage flappy test error

+ ApacheDS has old school testing lib which doesn't
  play nice with surefire.

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 - Fixing jetty-maven-plugin exclusions for demos

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 - More post move to /demos/ updates

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Merge remote-tracking branch 'origin/jetty-10.0.x' into jetty-10.0.x-5264-demo-modules

joakime committed:
Issue #5264 - Fixing test-integration references to demos

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 - Javadoc processing excludes for demos

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 - Adding HostHeaderCustomizer to H2/H2C Transport tests

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 - More javadoc skippedModules

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 -Fixing name

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 - making embedded examples more reliable

+ Discovering ${jetty.version} better
+ Creating ${jetty.base}/logs/ on JettyDemoBase usage

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 - identifying optional deps in jetty-home

 + Done to ease dependency / exclusions listing requirements
   elsewhere in the reactor.
   (Prevents accidental discovered components from
    triggering actions during use of jetty-home
    as a dependency)

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 - Fixing OSGi tests

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 - Fixing demo-async-rest module reference

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 - Fixing OSGi symbolic name references

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5264 - Addressing (some) aggregate-javadoc issues

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>
